### PR TITLE
Custom exceptions

### LINKED ISSUE #180: Custom Exceptions
I think it would be nice to come up with a set of custom exceptions that we consistently use for InterMol specific problems. That way we can hopefully provide more useful output and ensure that our testing summaries are formatted nicely. 

#177 introduces `MultipleValidationErrors` as a convenience for the final output summary.

On my wish list are:
* `UnsupportedConversion`, for when users attempt A->B and a functional form is supported in package A but not B. This should accept as its arguments the unsupported functional and the program that doesn't support it.
* `UnimplementedConversion`, for things we intend to support but haven't implemented. Could have an optional argument referencing a github issue.
* `ParsingError`, for issues during reading

### PR DESCRIPTION
**Pull request by ctk3b**
Resolves #180

### CODE REVIEW

**Comment by ctk3b** [comment]
This should be good to merge if you're ok with the changes, @mrshirts.

File: intermol/lammps/lammps_parser.py [hunk 1]
```diff
@@ -198,12 +200,12 @@ def canonical_dihedral(self, params, dihedral, direction='into'):
...
+            elif dihedral == ImproperHarmonicDihedral:
                 params['k'] *= canonical_force_scale
                 typename = 'harmonic'
                 paramlist = [params]
             else:
-                raise ValueError('A non-canonical dihedral was found in the '
+                raise InterMolError('A non-canonical dihedral was found in the '
                                  'system. All dihedrals should have been '
                                  'converted to either TrigDihedralType or '
                                  'ImproperHarmonicType so something likely '
```
Comment (mrshirts) [inline intermol/lammps/lammps_parser.py]: I changed these a bit in my commit.  Will have to merge later, but I think it's more general.

File: intermol/lammps/lammps_parser.py [hunk 2]
```diff
@@ -534,9 +536,10 @@ def parse_pair_modify(self, line):
             elif line[2] == 'arithmetic':
                 self.system.combination_rule = 'Lorentz-Berthelot'
             else:
-                raise ValueError('Unsupported pair_modify mix argument in input file!')
+                raise LammpsError('pair_modify mix argument "{}" is not yet supported'
+                                  'in InterMol'.format(line[2]))
         else:
-            raise ValueError('Unsupported pair_modify style in input file!')
+            raise LammpsError('pair_modify style not yet supported in InterMol')
```
Comment (mrshirts) [inline intermol/lammps/lammps_parser.py]: Is this an Lammps error or an Unimplemented error?

File: intermol/lammps/lammps_parser.py [hunk 3]
```diff
@@ -498,21 +500,21 @@ def parse_dimension(self, line):
...
+            raise LammpsError("Boundaries do not match specified dimension "
                              "in input file")
 
     def parse_pair_style(self, line):
         """ """
         self.pair_style = []
         if line[1] == 'hybrid':
-            logger.warning('Hybrid pair styles not yet implemented.')
+            raise LammpsError('Hybrid pair styles not yet implemented in InterMol.')
         elif line[1] in ('lj/cut/coul/long', 'lj/cut'):
```
Comment (mrshirts) [inline intermol/lammps/lammps_parser.py]: Lammps error or Unimeplented?

File: intermol/tests/lammps/unit_tests/atom_style-bond/log.lammps [hunk 4]
```diff
@@ -12,7 +12,7 @@ bond_style hybrid harmonic morse
 special_bonds lj 0.0 0.0 0.5 coul 0.0 0.0 0.5
```
Comment (mrshirts) [inline intermol/tests/lammps/unit_tests/atom_style-bond/log.lammps]: Shoudl this file be in the repository?

File: intermol/lammps/lammps_parser.py (hunk 2 above)
Comment (ctk3b) [inline intermol/lammps/lammps_parser.py]: Hmm right now I've only added `UnimplementedConversion` which is more specific for functional forms. I guess I should generalize it a bit more to work for any unimplemented feature.

File: intermol/tests/lammps/unit_tests/atom_style-bond/log.lammps (hunk 4 above)
Comment (ctk3b) [inline intermol/tests/lammps/unit_tests/atom_style-bond/log.lammps]: Fixed. Somehow a select few got added...

**Comment by ctk3b** [comment]
I've added some more general exceptions for unimplemented/unsupported settings. Let me know what you think.

File: intermol/desmond/desmond_parser.py [hunk 5]
```diff
@@ -7,6 +7,8 @@
 from intermol.atom import Atom
 from intermol.forces import *
 import intermol.forces.forcefunctions as ff
+from intermol.exceptions import (UnimplementedFunctional, UnsupportedFunctional,
+                                 DesmondError, InterMolError)
```
Comment (mrshirts) [inline intermol/desmond/desmond_parser.py]: Should UnsupportedSetting be here too?

File: intermol/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+class UnsupportedSetting(ConversionError):
+    """"""
+    def __str__(self):
+        return "{} is not supported in {}.".format(
+            self.could_not_convert, self.engine)
+
+
+class UnimplementedSetting(ConversionError):
+    """"""
```
Comment (mrshirts) [inline intermol/exceptions.py]: Maybe a bit of documentation here about when to use what situation - Setting vs Functional?  Might not be clear to someone else coming in (not entirely clear to me . . . )

File: intermol/desmond/desmond_parser.py (hunk 5 above)
Comment (ctk3b) [inline intermol/desmond/desmond_parser.py]: I've added them but they're currently unused.

File: intermol/exceptions.py (hunk 6 above)
Comment (ctk3b) [inline intermol/exceptions.py]: I've added some doc strings. I agree that it's not the most inuitive naming scheme right now. Any suggestions?

**Comment by ctk3b** [comment]
@mrshirts, are you ok with these changes? Any suggestions for improving the naming of the exceptions?

**Comment by mrshirts** [comment]
Sounds good, I was just in meetings for the last 4 hours.

**Comment by mrshirts** [comment]
Actually, the apostrophes are a bit ungrammatical, i.e.

    BendingTorsionDihedral's are not supported in desmond.

Maybe 

The BendingTorsionDihedral term is not supported in desmond?

Also, it should be Desmond, not desmond.

gromacs should the GROMACS (going by official decision)

lammps should be LAMMPS (going by website).

**Pull request merged**